---

_You are currently looking at **version 1.2** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [1]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [2]:
P1_Graphs = pickle.load(open('A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [3]:
# unique_degree_values = sorted(set(P1_Graphs[0].degree().values()))
# unique_degree_values

In [4]:
# [list(P1_Graphs[0].degree().values()).count(degree) for degree in unique_degree_values]

In [5]:
import numpy as np
for G in P1_Graphs:
    print('--------------')
    print('avg shortest path:', nx.average_shortest_path_length(G))
    print('avg clustering:', nx.average_clustering(G))
    unique_degree_values = sorted(set(G.degree().values()))
    print([list(G.degree().values()).count(degree) for degree in unique_degree_values])
#     print([degree_value.count(i) for i in degree_values])
    print('--------------')

--------------
avg shortest path: 4.099161161161161
avg clustering: 0.03167539146454044
[465, 237, 100, 56, 32, 26, 16, 14, 9, 7, 5, 5, 2, 1, 1, 2, 2, 3, 2, 1, 2, 2, 2, 1, 2, 1, 2, 1, 1]
--------------
--------------
avg shortest path: 5.089871871871872
avg clustering: 0.5642419635919628
[2, 23, 166, 615, 171, 22, 1]
--------------
--------------
avg shortest path: 9.378702269692925
avg clustering: 0.4018222222222227
[2, 93, 567, 79, 9]
--------------
--------------
avg shortest path: 3.1048046283934134
avg clustering: 0.03780379975223251
[259, 133, 89, 58, 47, 25, 19, 13, 15, 14, 8, 7, 5, 9, 4, 1, 1, 4, 6, 1, 2, 1, 6, 3, 1, 2, 2, 1, 2, 2, 3, 2, 1, 1, 1, 1, 1]
--------------
--------------
avg shortest path: 5.0785509568313305
avg clustering: 0.0033037037037037037
[102, 195, 218, 126, 74, 20, 12, 2, 1]
--------------


In [6]:
def graph_identification():
    
    # Your Code Here
    labels = []
    
    for G in P1_Graphs:
        avg_path = nx.average_shortest_path_length(G)
        avg_cluster = nx.average_clustering(G)
        unique_degree_values = sorted(set(G.degree().values()))
        deg_dist_len = len([list(G.degree().values()).count(degree) for degree in unique_degree_values])
        
        if deg_dist_len > 10:
            labels.append('PA')
        elif avg_cluster < 0.4:
            labels.append('SW_H')
        else:
            labels.append('SW_L')

    
    return labels# Your Answer Here

In [7]:
graph_identification()

['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']

---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagementSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagementSalary` indicates whether that person is receiving a management position salary.

In [8]:
G = nx.read_gpickle('email_prediction.txt')

print(nx.info(G))

Name: 
Type: Graph
Number of nodes: 1005
Number of edges: 16706
Average degree:  33.2458


### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a management position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a management salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a management position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 252 with the data being the probability of receiving management salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

In [9]:
# df = pd.DataFrame(index = G.nodes())
# df['ManagementSalary'] = pd.Series(nx.get_node_attributes(G, 'ManagementSalary'))
# df['Department'] = pd.Series(nx.get_node_attributes(G, 'Department'))
# df['Clustering'] = pd.Series(nx.clustering(G))
# df['Degree'] = pd.Series(nx.degree(G))
# df['degree_centrality'] = pd.Series(nx.degree_centrality(G))
# df['closeness'] = pd.Series(nx.closeness_centrality(G, normalized=True))
# df['betweeness'] = pd.Series(nx.betweenness_centrality(G, normalized=True))
# df['pr'] = pd.Series(nx.pagerank(G))
# df['auth'] = pd.Series(nx.hits(G)[0])
# df['hub'] = pd.Series(nx.hits(G)[1])

# df.head()

In [10]:
# pred_df = df[df['ManagementSalary'].isnull()]
# print(pred_df.shape)
# pred_df.head()

In [11]:
# dev_df = df[~df['ManagementSalary'].isnull()]
# print(dev_df.shape)
# dev_df.head()

In [12]:
# from sklearn.model_selection import train_test_split
# from sklearn.ensemble import RandomForestClassifier

In [13]:
# X = dev_df.drop('ManagementSalary', axis = 1)
# X.head()

In [14]:
# y = dev_df['ManagementSalary']
# y.head()

In [15]:
# X_train, X_test, y_train, y_test = train_test_split(X, y, stratify = y)

In [16]:
# RFC = RandomForestClassifier().fit(X, y)

In [17]:
# RFC.score(X_test, y_test)

In [18]:
# RFC.classes_

In [19]:
# X_pred = pred_df.drop('ManagementSalary', axis = 1)

In [20]:
# RFC.predict_proba(X_pred)[:,1]

In [21]:
# pred_df['scores'] = RFC.predict_proba(X_pred)[:,1]
# pred_df.head()

In [22]:
# pred_df['scores']

In [23]:
def salary_predictions():
    
    # Your Code Here
#     from sklearn.ensemble import RandomForestRegressor
    from sklearn.ensemble import RandomForestClassifier
    
    #create attribute matrix
    df = pd.DataFrame(index = G.nodes())
    df['ManagementSalary'] = pd.Series(nx.get_node_attributes(G, 'ManagementSalary'))
    df['Department'] = pd.Series(nx.get_node_attributes(G, 'Department'))
    df['Clustering'] = pd.Series(nx.clustering(G))
    df['Degree'] = pd.Series(nx.degree(G))
    df['degree_centrality'] = pd.Series(nx.degree_centrality(G))
    df['closeness'] = pd.Series(nx.closeness_centrality(G, normalized=True))
    df['betweeness'] = pd.Series(nx.betweenness_centrality(G, normalized=True))
    df['pr'] = pd.Series(nx.pagerank(G))
    df['auth'] = pd.Series(nx.hits(G)[0])
    df['hub'] = pd.Series(nx.hits(G)[1])
    
    #create df for nodes we need to predict
    pred_df = df[df['ManagementSalary'].isnull()]
    
    #create df for nodes to train model
    dev_df = df[~df['ManagementSalary'].isnull()]
    
    #define X and y
    X = dev_df.drop('ManagementSalary', axis = 1)
    y = dev_df['ManagementSalary']
    
    #create and fit classifier
#     RFC = RandomForestRegressor().fit(X, y)
    RFC = RandomForestClassifier().fit(X, y)
    
    #define nodes to predict
    X_pred = pred_df.drop('ManagementSalary', axis = 1)
    
    #predict scores
#     pred_df['scores'] = RFC.predict(X_pred)
    pred_df['scores'] = RFC.predict_proba(X_pred)[:,1]

    return pred_df['scores']# Your Answer Here

In [24]:
salary_predictions()

/opt/conda/lib/python3.6/site-packages/ipykernel/__main__.py:39: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


1       0.0
2       1.0
5       1.0
8       0.1
14      0.1
18      0.1
27      0.1
30      0.6
31      0.3
34      0.0
37      0.0
40      0.0
45      0.0
54      0.5
55      0.5
60      0.3
62      1.0
65      1.0
77      0.0
79      0.3
97      0.0
101     0.0
103     0.4
108     0.1
113     0.0
122     0.0
141     0.1
142     1.0
144     0.0
145     0.7
       ... 
913     0.0
914     0.0
915     0.0
918     0.1
923     0.0
926     0.0
931     0.0
934     0.0
939     0.0
944     0.0
945     0.0
947     0.1
950     0.1
951     0.0
953     0.0
959     0.0
962     0.0
963     0.3
968     0.0
969     0.0
974     0.0
984     0.0
987     0.0
989     0.0
991     0.0
992     0.0
994     0.0
996     0.0
1000    0.0
1001    0.0
Name: scores, Length: 252, dtype: float64

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [25]:
future_connections = pd.read_csv('Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [26]:
# df = pd.DataFrame(index = G.edges())
# df.head()

In [27]:
# df = df.merge(future_connections, how = 'outer', left_index= True, right_index= True)
# df.head()

In [28]:
# df[~df['Future Connection'].isnull()]

In [29]:
# df = pd.DataFrame(index = G.edges())
# df = df.merge(future_connections, how = 'outer', left_index= True, right_index= True)
# df['preferential attachment'] = [i[2] for i in nx.preferential_attachment(G, df.index)]
# df['Common Neighbors'] = df.index.map(lambda nodes: len(list(nx.common_neighbors(G, nodes[0], nodes[1]))))
# df.head()

In [30]:
# list(nx.jaccard_coefficient(G))

In [31]:
# df['jaccard'] = [value[2] for value in nx.jaccard_coefficient(G, df.index)]
# df.head()

In [32]:
# df['rescource_alloc'] = [value[2] for value in nx.resource_allocation_index(G, df.index)]
# df.head()

In [33]:
# for name in G.nodes():
#     G.node[name]['community'] = G.node[name]['Department']

In [34]:
# df['soundarajan'] = [value[2] for value in nx.cn_soundarajan_hopcroft(G, df.index)]
# df.head()

In [35]:
# pred_df = df[df['Future Connection'].isnull()]
# pred_df.head()

In [36]:
# #create df for nodes to train model
# dev_df = df[~df['Future Connection'].isnull()]
# dev_df.head()

In [37]:
# #define X and y
# X = dev_df.drop('Future Connection', axis = 1)
# y = dev_df['Future Connection']

In [38]:
# #create and fit classifier
# from sklearn.ensemble import RandomForestClassifier
# RFC = RandomForestClassifier().fit(X, y)

In [39]:
# #define nodes to predict
# X_pred = pred_df.drop('Future Connection', axis = 1)

In [40]:
# #predict scores
# pred_df['scores'] = RFC.predict_proba(X_pred)[:,1]
# pred_df['scores']

In [ ]:
def new_connections_predictions():
    
    # Your Code Here
    df = pd.DataFrame(index = G.edges())
    df = df.merge(future_connections, how = 'outer', left_index= True, right_index= True)
    df['preferential attachment'] = [i[2] for i in nx.preferential_attachment(G, df.index)]
    df['Common Neighbors'] = df.index.map(lambda nodes: len(list(nx.common_neighbors(G, nodes[0], nodes[1]))))
    df['jaccard'] = [value[2] for value in nx.jaccard_coefficient(G, df.index)]
    df['rescource_alloc'] = [value[2] for value in nx.resource_allocation_index(G, df.index)]
    
    for name in G.nodes():
        G.node[name]['community'] = G.node[name]['Department']
        
    df['soundarajan'] = [value[2] for value in nx.cn_soundarajan_hopcroft(G, df.index)]

    pred_df = df[df['Future Connection'].isnull()]
    
    dev_df = df[~df['Future Connection'].isnull()]

    #define X and y
    X = dev_df.drop('Future Connection', axis = 1)
    y = dev_df['Future Connection']
    
    #create and fit classifier
    from sklearn.ensemble import RandomForestClassifier
    RFC = RandomForestClassifier().fit(X, y)
    
    #define nodes to predict
    X_pred = pred_df.drop('Future Connection', axis = 1)
    
    #predict scores
    pred_df['scores'] = RFC.predict_proba(X_pred)[:,1]
    
    return pred_df['scores']# Your Answer Here

In [ ]:
new_connections_predictions()